### PR TITLE
Optimize / Cleanup QueryStringDecoder / QueryStringEncoder

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

QueryStringDecoder and QueryStringEncoder contained some code that could either cleaned-up or optimized.

Modifications:

- Fix typos in exception messages and javadocs
- Precompile Pattern
- Make use of StringUtil.EMPTY_STRING

Result:

Faster and cleaner code.

### CODE REVIEW

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java [hunk 1]
```diff
@@ -140,26 +137,19 @@ public QueryStringDecoder(URI uri, Charset charset) {
...
         if (rawPath != null) {
             hasPath = true;
         } else {
-            rawPath = "";
+            rawPath = EMPTY_STRING;
             hasPath = false;
         }
         // Also take care of cut of things like "http://localhost"
-        this.uri = rawPath + (uri.getRawQuery() == null? "" : '?' + uri.getRawQuery());
+        this.uri = rawPath + (uri.getRawQuery() == null? EMPTY_STRING : '?' + uri.getRawQuery());
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java]: Isn't there a useless intermediate concat when hitting the second branch?

Would this be better?
`uri.getRawQuery() == null? rawPath : rawPath + "?" + uri.getRawQuery()`

Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringDecoder.java]: +1

**Comment by CodingFabian** [comment]
lgtm

File: codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java [hunk 2]
```diff
@@ -37,6 +40,7 @@
  */
 public class QueryStringEncoder {
 
+    private static final Pattern PATTERN = Pattern.compile("\\+");
```
Comment (trustin) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java]: Not sure which is better, but how about:

```java
Pattern.compile("+", Pattern.LITERAL)
```

Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java]: yeah let us do this :)

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (d675febf07d14d4dff82471829f974369705655a) and 4.0 (67224a8da1c528b8f6f5343f73fbc24733bc9f78)

**Pull request closed without merging**